File: bindings/python/insight/signal/acoustics.py
"""Acoustic frequency scale functions for signal processing.

Provides conversions between Hz, Mel, and Bark frequency scales,
and Mel frequency bank generation.
"""

from insight._insight import signal as _signal

__all__ = [
    "mel2hz",
    "hz2mel",
    "hz2bark",
    "bark2hz",
    "mel_frequencies",
]


def mel2hz(mel):
    """Convert Mel-scale frequencies to Hz.

    Args:
        mel: Mel-scale frequency value or array.

    Returns:
        Corresponding frequency in Hz.
    """
    return _signal.mel2hz(mel)


def hz2mel(hz):
    """Convert frequencies in Hz to the Mel scale.

    Args:
        hz: Frequency value or array in Hz.

    Returns:
        Corresponding Mel-scale value.
    """
    return _signal.hz2mel(hz)


def hz2bark(hz):
    """Convert frequencies in Hz to the Bark scale.

    Args:
        hz: Frequency value or array in Hz.

    Returns:
        Corresponding Bark-scale value.
    """
    return _signal.hz2bark(hz)


def bark2hz(bark):
    """Convert Bark-scale frequencies to Hz.

    Args:
        bark: Bark-scale frequency value or array.

    Returns:
        Corresponding frequency in Hz.
    """
    return _signal.bark2hz(bark)


def mel_frequencies(num_mel=128, fmin=0.0, fmax=11025.0, htk=False):
    """Compute Mel-spaced frequencies.

    Args:
        num_mel: Number of Mel frequencies to generate.  Default is 128.
        fmin: Minimum frequency in Hz.  Default is 0.0.
        fmax: Maximum frequency in Hz.  Default is 11025.0.
        htk: If True, use the HTK formula instead of Slaney's.  Default is
            False.

    Returns:
        1-D array of *num_mel* equally spaced frequencies on the Mel scale.

    Note:
        The *htk* parameter is accepted for API compatibility but is
        currently ignored by the backend.
    """
    return _signal.mel_frequencies(num_mel, f_low=fmin, f_high=fmax)

File: bindings/lua/insight/signal/acoustics.lua
--- Acoustic scale functions.
-- Provides conversions between Hz, Mel, and Bark frequency scales.
-- @module insight.signal.acoustics

local native = require("_insight")
local sig = native.signal

local M = {}

local _wrap = require("insight._wrap")

local from_table = native.from_table

--- Convert Mel scale to Hz.
-- Accepts a scalar number or an Array.
-- @param mel Frequency in Mel scale (number or Array).
-- @treturn number Frequency in Hz (scalar).
M.mel2hz = _wrap({ "mel" }, function(mel)
  local arr = (type(mel) == "number") and from_table({ mel }) or mel
  local r = sig.mel2hz(arr)
  if type(mel) == "number" then
    return r:item(0)
  end
  return r
end)

--- Convert Hz to Mel scale.
-- Accepts a scalar number or an Array.
-- @param hz Frequency in Hz (number or Array).
-- @treturn number Frequency in Mel scale (scalar).
M.hz2mel = _wrap({ "hz" }, function(hz)
  local arr = (type(hz) == "number") and from_table({ hz }) or hz
  local r = sig.hz2mel(arr)
  if type(hz) == "number" then
    return r:item(0)
  end
  return r
end)

--- Compute Mel-spaced frequencies.
-- @int num Number of frequencies to generate.
-- @number[opt=0] fmin Minimum frequency in Hz.
-- @number[opt] fmax Maximum frequency in Hz (default: fs/2).
-- @bool[opt=false] htk Use HTK formula.
-- @treturn Array Mel-spaced frequencies.
M.mel_frequencies = _wrap({ "num", "fmin", "fmax", "htk" }, function(num, fmin, fmax, htk)
  return sig.mel_frequencies(num, fmin or 0, fmax or 8000, htk or false)
end)

--- Convert Hz to Bark scale.
-- Accepts a scalar number or an Array.
-- @param hz Frequency in Hz (number or Array).
-- @treturn number Frequency in Bark scale (scalar).
M.hz2bark = _wrap({ "hz" }, function(hz)
  local arr = (type(hz) == "number") and from_table({ hz }) or hz
  local r = sig.hz2bark(arr)
  if type(hz) == "number" then
    return r:item(0)
  end
  return r
end)

--- Convert Bark scale to Hz.
-- Accepts a scalar number or an Array.
-- @param bark Frequency in Bark scale (number or Array).
-- @treturn number Frequency in Hz (scalar).
M.bark2hz = _wrap({ "bark" }, function(bark)
  local arr = (type(bark) == "number") and from_table({ bark }) or bark
  local r = sig.bark2hz(arr)
  if type(bark) == "number" then
    return r:item(0)
  end
  return r
end)

return M
